consensus/view_change.go

```diff
@@ -214,228 +188,73 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
+		return
 	}
 
-	preparedBlock := &types.Block{}
-	hasBlock := false
-	if len(recvMsg.Payload) != 0 && len(recvMsg.Block) != 0 {
-		if err := rlp.DecodeBytes(recvMsg.Block, preparedBlock); err != nil {
-			consensus.getLogger().Warn().
-				Err(err).
-				Uint64("MsgBlockNum", recvMsg.BlockNum).
-				Msg("[onViewChange] Unparseable prepared block data")
-			return
-		}
-		hasBlock = true
+	msgType, preparedBlock, err := consensus.vc.VerifyViewChangeMsg(recvMsg, members)
+	if err != nil {
+		consensus.getLogger().Error().Err(err).
+			Uint64("viewID", recvMsg.ViewID).
+			Uint64("blockNum", recvMsg.BlockNum).
+			Str("msgSender", recvMsg.SenderPubkey.Bytes.Hex()).
+			Msg("Verify View Change Message Error")
+		return
 	}
 
-	// m2 type message
-	if !hasBlock {
-		_, ok := consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M2 message from validator")
-			return
-		}
-
-		if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, NIL) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed To Verify Signature For M2 Type Viewchange Message")
-			return
-		}
-
-		consensus.getLogger().Info().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Add M2 (NIL) type message")
-		consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewchangeSig
-		consensus.nilBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true) // Set the bitmap indicating that this validator signed.
-	} else { // m1 type message
-		if consensus.BlockVerifier(preparedBlock); err != nil {
-			consensus.getLogger().Error().Err(err).Msg("[onViewChange] Prepared block verification failed")
-			return
-		}
-		_, ok := consensus.bhpSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M1 Message From the Validator")
-			return
-		}
-		if !recvMsg.ViewchangeSig.VerifyHash(recvMsg.SenderPubkey.Object, recvMsg.Payload) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed to Verify Signature for M1 Type Viewchange Message")
-			return
-		}
-
-		// first time receive m1 type message, need verify validity of prepared message
-		if len(consensus.m1Payload) == 0 || !bytes.Equal(consensus.m1Payload, recvMsg.Payload) {
-			if len(recvMsg.Payload) <= 32 {
-				consensus.getLogger().Warn().
-					Int("len", len(recvMsg.Payload)).
-					Msg("[onViewChange] M1 RecvMsg Payload Not Enough Length")
-				return
-			}
-			blockHash := recvMsg.Payload[:32]
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(recvMsg.Payload, 32)
-			if err != nil {
-				consensus.getLogger().Error().Err(err).Msg("[onViewChange] M1 RecvMsg Payload Read Error")
-				return
-			}
-
-			if !consensus.Decider.IsQuorumAchievedByMask(mask) {
```

Comment:
Seems the quorum check on the prepared msg is lost in the new code. Please double check.

Reply:
Quorum check is in line 241 of the new file. Always use m3 to check quorum, as we always add m3 sig and bitmap.  Does it make sense?  The original code checked quorum on m1payload, which is not necessary, IMO.

Reply:
It's not about the quorum on m1 view change message. This is about the quorum on the prepared signatures in m1. We definitely need to check the prepared signature reached quorum on the block before trusting it.